main.py

```diff
@@ -1,16 +1,416 @@
+        # [df.loc[:, ~df.columns.isin(['Vineyard', 'Treatment', 'Replication', 'Time','new_time'])], df_out.add_suffix('_out')],
+        axis=1)
+    correl_out = results.corr()
+    out_cols = [col for col in correl_out.columns if '_out' in col]
+    correl_out = correl_out[out_cols]
+    correl_out = correl_out[~correl_out.index.str.contains('_out')]
+    plot_correl_matrix(correl_out,correlation_name='in_vs_out')
+
+
+def get_storage_per_treatment(storage_df: dict, exp_desc: dict):
+    '''
+    :param storage_df: dictionary contains all storage data, key stands for the treatment
+    :param exp_desc: dictionary contains all storage relevant data per treatment
+    :return: dictionary contains log of storage data including interuptions.
+    '''
+    storage_df_per_treatment = {}
+    for treatment, values in exp_desc.items():
+        mask_first_period = storage_df[values[0]]['weeks'] <= values[1]
+        first_period = storage_df[values[0]].loc[mask_first_period]
+        mask_second_period = storage_df[values[2]]['weeks'] > values[1]
+        second_period = storage_df[values[2]].loc[mask_second_period]
+        storage_df_per_treatment[treatment] = pd.concat([first_period, second_period])
+    return storage_df_per_treatment
+
+
+def enrich_storage_per_treatment(storage_dict: dict):
+    '''
+    :param storage_dict: receive dict with log of the storage data
+    :return: dictionary contains some statistics on log of the per each treatment
+    '''
+    for treatment, df_treatment_storage in storage_dict.items():
+        df_treatment_storage['Humidity_cumsum'] = df_treatment_storage['Humidity'].cumsum()
+        df_treatment_storage['Temperature_cumsum'] = df_treatment_storage['Temperature'].cumsum()
+        df_treatment_storage['Humidity_max'] = df_treatment_storage['Humidity'].max()
+        df_treatment_storage['Temperature_max'] = df_treatment_storage['Temperature'].max()
+        df_treatment_storage['Temperature_max_relative'] = (
+                df_treatment_storage['Temperature'] - df_treatment_storage['room_temp']).max()
+        df_treatment_storage['mapped_period'] = df_treatment_storage.apply(map_weeks, axis=1)
+        storage_dict[treatment] = df_treatment_storage
+    return storage_dict
+
+
+def map_weeks(row):
+    '''
+    :param row: mapping function from weeks to period in test
+    :return: period in test, numbered 1-4
+    '''
+    if row['weeks'] <= 3:
+        return 1
+    elif row['weeks'] <= 6:
+        return 2
+    elif row['weeks'] <= 9:
+        return 3
+    elif row['weeks'] <= 12:
+        return 4
+    else:
+        return 9999
+
+
+def get_relevant_data_per_period(storage_dict: dict):
+    '''
+    :param storage_dict: dictionary contians storage data
+    :return: dict contains statistics per each period per treatment
+    '''
+    final_storage_data_per_treatment = {}
+    data_storage_for_all_periods = {}
+    for treatments, values in storage_dict.items():
```

Comment:
can be done without loop

Reply:
how the hell? 
how you can run over dict values w/o loop?